cli/cache/check.rs

```diff
@@ -0,0 +1,213 @@
+use std::path::Path;
+
+use deno_ast::ModuleSpecifier;
+use deno_core::anyhow::Context;
+use deno_core::error::AnyError;
+use deno_runtime::deno_webstorage::rusqlite::params;
+use deno_runtime::deno_webstorage::rusqlite::Connection;
+
+use super::common::run_sqlite_pragma;
+
+/// The cache used to tell whether type checking should occur again.
+///
+/// This simply stores a hash of the inputs of each successful type check
+/// and only clears them out when changing CLI versions.
+pub struct TypeCheckCache {
+  conn: Connection,
+}
+
+impl TypeCheckCache {
+  pub fn new(db_file_path: &Path) -> Result<Self, AnyError> {
```

Comment:
meta: we have never managed the sizes of caches, and having an unbounded database it raises the question of quota management and how we would manage an LRU and size constraints.

Reply:
I didn't think it was worth the extra complexity, but maybe it is? The previous tsbuildinfo cache was also unbounded.

I just wrote some tests for this and for the same entrypoint type checked multiple times with a unique change each time, the end result was this (the hashes don't take up a lot of room):

* 1,000 unique type checks - 92KB
* 5,000 unique type checks - 316KB
* 10,000 unique type checks - 600KB

Also, I didn't notice any slowdown in speed as the database size grew.

However, once you start type checking different root entry points, then it starts to grow a lot because it stores the tsbuildinfo as well. For example, this is with unique entrypoints:

* 1,000 unique entry points - 6,250KB
* 5,000 unique entry points - ~30MB
* 10,000 unique entry points - 59.5MB

For this, I also didn't notice any slowdown as the database size grew... it remained exactly the same. SQLite seems quite fast.

I think it would be unlikely for someone to even have more than 100 root entrypoints before upgrading their CLI version. Let me know your thoughts.

Reply:
Yeah, I don't disagree, it was more of a meta concern that this plus other stuff likely needs better quota management.